### PR TITLE
Syncing IPVS conntrack cleaning with IPTables.

### PR DESCRIPTION
**Pull request by aroradaman**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by daman1807** [comment]
@aojea created this PR for syncing conntrack cleaning ..

**Comment by daman1807** [comment]
@aojea I think gci-gce-ipvs is flaking (not a permanent failure).
It passed at second attempt.

**Comment by aojea** [comment]
> @aojea I think gci-gce-ipvs is flaking (not a permanent failure).

iptables does not flake ;) 

it means that probably the connection tracking is not at the right time?

```
TEP: checking client pod connected to the backend 2 on Node bootstrap-e2e-minion-group-zmrs - test/e2e/network/loadbalancer.go:1273 @ 03/01/23 11:43:57.072
Mar  1 11:44:00.093: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:04.094: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:08.096: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:12.097: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:16.098: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:20.099: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:24.101: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:28.102: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:32.103: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:36.105: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:40.106: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:44.108: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:48.109: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:52.111: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:44:56.113: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:00.115: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:04.117: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:08.118: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:12.119: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:16.120: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:20.121: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:24.123: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:28.124: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:32.125: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:36.127: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:40.128: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:44.129: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:48.131: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:52.133: INFO: Connected successfully to: 34.83.42.133:80
Mar  1 11:45:56.135: INFO: Connected successfully to: 34.83.42.133:80
[FAILED] Failed to connect to backend 2
In [It] at: test/e2e/network/loadbalancer.go:1279 @ 03/01/23 11:45:57.073
< Exit [It] should be able to preserve UDP traffic when server pod cycles for a Lo
```

**Comment by daman1807** [comment]
@aojea this will be reproducible in local kind/metallb setup?
will try to dig deep into it

**Comment by aojea** [comment]
> @aojea this will be reproducible in local kind/metallb setup?

I don't know honestly, the metalb loadbalancer doesn't take into account healtchecks per example AFAIK

**Comment by aojea** [comment]
but I'm fine consolidating the code, merging and iterating ....

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -955,8 +956,13 @@ func (proxier *Proxier) syncProxyRules() {
 			for _, extIP := range svcInfo.ExternalIPStrings() {
 				staleServices.Insert(extIP)
 			}
-			for _, extIP := range svcInfo.LoadBalancerIPStrings() {
-				staleServices.Insert(extIP)
+			for _, lbIP := range svcInfo.LoadBalancerIPStrings() {
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:959]: can you replace the code with exactly the same code and comments than in kube-proxy iptables?

```
	serviceUpdateResult := proxier.svcPortMap.Update(proxier.serviceChanges)
	endpointUpdateResult := proxier.endpointsMap.Update(proxier.endpointsChanges)

	// We need to detect stale connections to UDP Services so we
	// can clean dangling conntrack entries that can blackhole traffic.
	conntrackCleanupServiceIPs := serviceUpdateResult.UDPStaleClusterIP
	conntrackCleanupServiceNodePorts := sets.NewInt()
	// merge stale services gathered from updateEndpointsMap
	// an UDP service that changes from 0 to non-0 endpoints is considered stale.
	for _, svcPortName := range endpointUpdateResult.StaleServiceNames {
		if svcInfo, ok := proxier.svcPortMap[svcPortName]; ok && svcInfo != nil && conntrack.IsClearConntrackNeeded(svcInfo.Protocol()) {
			klog.V(2).InfoS("Stale service", "protocol", strings.ToLower(string(svcInfo.Protocol())), "servicePortName", svcPortName, "clusterIP", svcInfo.ClusterIP())
			conntrackCleanupServiceIPs.Insert(svcInfo.ClusterIP().String())
			for _, extIP := range svcInfo.ExternalIPStrings() {
				conntrackCleanupServiceIPs.Insert(extIP)
			}
			for _, lbIP := range svcInfo.LoadBalancerIPStrings() {
				conntrackCleanupServiceIPs.Insert(lbIP)
			}
			nodePort := svcInfo.NodePort()
			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
				klog.V(2).InfoS("Stale service", "protocol", strings.ToLower(string(svcInfo.Protocol())), "servicePortName", svcPortName, "nodePort", nodePort)
				conntrackCleanupServiceNodePorts.Insert(nodePort)
			}
		}
	}

```

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1061,11 +1067,11 @@ func (proxier *Proxier) syncProxyRules() {
 			klog.ErrorS(nil, "Failed to cast serviceInfo", "servicePortName", svcPortName)
 			continue
 		}
-		isIPv6 := netutils.IsIPv6(svcInfo.ClusterIP())
-		localPortIPFamily := netutils.IPv4
-		if isIPv6 {
-			localPortIPFamily = netutils.IPv6
-		}
+		//isIPv6 := netutils.IsIPv6(svcInfo.ClusterIP())
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1070]: this is for testing no? we should remove these lines

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -1554,6 +1560,14 @@ func (proxier *Proxier) syncProxyRules() {
 			klog.ErrorS(err, "Failed to delete stale service IP connections", "IP", svcIP)
 		}
 	}
+
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1563]: these are the ones in iptables
```
	klog.V(4).InfoS("Deleting conntrack stale entries for services", "IPs", conntrackCleanupServiceIPs.UnsortedList())
	for _, svcIP := range conntrackCleanupServiceIPs.UnsortedList() {
		if err := conntrack.ClearEntriesForIP(proxier.exec, svcIP, v1.ProtocolUDP); err != nil {
			klog.ErrorS(err, "Failed to delete stale service connections", "IP", svcIP)
		}
	}
	klog.V(4).InfoS("Deleting conntrack stale entries for services", "nodePorts", conntrackCleanupServiceNodePorts.UnsortedList())
	for _, nodePort := range conntrackCleanupServiceNodePorts.UnsortedList() {
		err := conntrack.ClearEntriesForPort(proxier.exec, nodePort, isIPv6, v1.ProtocolUDP)
		if err != nil {
			klog.ErrorS(err, "Failed to clear udp conntrack", "nodePort", nodePort)
		}
	}
	klog.V(4).InfoS("Deleting stale endpoint connections", "endpoints", endpointUpdateResult.StaleEndpoints)
	proxier.deleteEndpointConnections(endpointUpdateResult.StaleEndpoints)
```

why doesn't match?

File: pkg/proxy/ipvs/proxier.go [hunk 4]
```diff
@@ -1554,6 +1560,14 @@ func (proxier *Proxier) syncProxyRules() {
...
+	klog.V(4).InfoS("Deleting conntrack stale entries for services", "nodePorts", conntrackCleanupServiceNodePorts.UnsortedList())
+	for _, nodePort := range conntrackCleanupServiceNodePorts.UnsortedList() {
+		err := conntrack.ClearEntriesForPort(proxier.exec, nodePort, proxier.ipFamily == v1.IPv6Protocol, v1.ProtocolUDP)
+		if err != nil {
+			klog.ErrorS(err, "Failed to clear udp conntrack", "nodePort", nodePort)
+		}
+	}
 	proxier.deleteEndpointConnections(endpointUpdateResult.StaleEndpoints)
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1571]: can you check this method has exactly the same code than the one in iptables?

File: pkg/proxy/ipvs/proxier.go [hunk 5]
```diff
@@ -1310,31 +1319,31 @@ func (proxier *Proxier) syncProxyRules() {
...
-				if utilproxy.IsZeroCIDR(address) {
-					// Empty IP address means all
-					lp.IP = ""
-					lps = append(lps, lp)
-					// If we encounter a zero CIDR, then there is no point in processing the rest of the addresses.
-					break
-				}
-				lps = append(lps, lp)
-			}
+			//var lps []netutils.LocalPort
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1322]: can you completely remove the lines? not commenting them?

File: pkg/proxy/ipvs/proxier.go [hunk 6]
```diff
@@ -1547,17 +1556,27 @@ func (proxier *Proxier) syncProxyRules() {
 		klog.ErrorS(err, "Error syncing healthcheck endpoints")
 	}
 
+	metrics.SyncProxyRulesNoLocalEndpointsTotal.WithLabelValues("internal").Set(float64(proxier.serviceNoLocalEndpointsInternal.Len()))
+	metrics.SyncProxyRulesNoLocalEndpointsTotal.WithLabelValues("external").Set(float64(proxier.serviceNoLocalEndpointsExternal.Len()))
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1560]: are these metrics related to this PR?

File: pkg/proxy/ipvs/proxier.go [hunk 7]
```diff
@@ -1547,17 +1556,27 @@ func (proxier *Proxier) syncProxyRules() {
...
+
 	// Finish housekeeping.
+	// Clear stale conntrack entries for UDP Services, this has to be done AFTER the iptables rules are programmed.
```
[the remark this answers is not in the document]
Comment (daman1807) [reply pkg/proxy/ipvs/proxier.go:1563]: lol, my bad :p

File: pkg/proxy/ipvs/proxier.go (hunk 5 above)
Comment (daman1807) [reply pkg/proxy/ipvs/proxier.go:1322]: okay

File: pkg/proxy/ipvs/proxier.go (hunk 6 above)
Comment (daman1807) [reply pkg/proxy/ipvs/proxier.go:1560]: no, but they are tracked before conntrack cleaning in IPtables, hence moved their position here in ipvs proxier.

Comment (aojea) [reply pkg/proxy/ipvs/proxier.go:1534]: then use independent commit for the metrics, if we want to backport the fix we may have conflicts with them

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (aojea) [reply pkg/proxy/ipvs/proxier.go:1073]: comment stands, please remote these lines instead of commenting

**Comment by aojea** [comment]
almost there

/kind bug

this solves a bug, it was really out of sync the conntrack logic

File: pkg/proxy/ipvs/proxier.go [hunk 8]
```diff
@@ -1825,27 +1814,36 @@ func (proxier *Proxier) createAndLinkKubeChain() {
 }
 
 // After a UDP or SCTP endpoint has been removed, we must flush any pending conntrack entries to it, or else we
-// risk sending more traffic to it, all of which will be lost (because UDP).
+// risk sending more traffic to it, all of which will be lost.
 // This assumes the proxier mutex is held
+// TODO: move it to util
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1819]: this is the first candidate to consolidate, only has the exec dependency, but better in a follow up

**Comment by daman1807** [comment]
/test pull-kubernetes-e2e-gci-gce-ipvs

cluster was not created, again.

**Comment by aojea** [comment]
> /test pull-kubernetes-e2e-gci-gce-ipvs
> 
> cluster was not created, again.

there is an issue in the CI that may be related, it is still unsolved AFAIK

**Comment by daman1807** [comment]
@aojea in the following snippet, do we consider 30sec for conntrack entry timeout at the load balancer?
https://github.com/kubernetes/kubernetes/blob/3835c7aecd0c55e6a10e9955347fee10889c368b/test/e2e/network/loadbalancer.go#L1242-L1248

I ran the failing test locally in kind, watched the conntrack table entries, and enabled verbose logging for kube-proxy.  The conntrack entries were being cleared up as per logic, but surprisingly in the contract table the 30s timeout was replaced by 120s timeout.

I later found out that when multiple requests and replies occur between the same socket pairs, the flow entry is considered to be a stream and conntrack starts using _nf_conntrack_udp_timeout_stream_ instead of _nf_conntrack_udp_timeout_.

And the timeout which we have in the failing test is exactly the same(120s), could this be the reason for the flaky behavior?



diff for conntrack table output(XML)
```
        <meta direction="independent">
-               <timeout>29</timeout>
+               <timeout>119</timeout>
                <mark>0</mark>
                <use>1</use>
                <id>3760020581</id>
+               <assured/>
        </meta>
```

**Comment by danwinship** [comment]
> I think that it is better if we merge this before any refactoring

this _is_ a refactoring!

File: pkg/proxy/ipvs/proxier.go [hunk 9]
```diff
@@ -1310,32 +1315,6 @@ func (proxier *Proxier) syncProxyRules() {
...
-				}
-				lps = append(lps, lp)
-			}
-
-			// For ports on node IPs, open the actual port and hold it.
-			for _, lp := range lps {
-				if svcInfo.Protocol() != v1.ProtocolSCTP && lp.Protocol == netutils.UDP {
-					conntrack.ClearEntriesForPort(proxier.exec, lp.Port, isIPv6, v1.ProtocolUDP)
-				}
-			}
```
Comment (danwinship) [inline pkg/proxy/ipvs/proxier.go:1337]: oof, I modified this code in #116016 and didn't even notice how utterly broken it was since the removal of the port opener stuff...
